packages/flutter_test/lib/src/controller.dart

```diff
@@ -24,6 +24,32 @@ const double kDragSlopDefault = 20.0;
 
 const String _defaultPlatform = kIsWeb ? 'web' : 'android';
 
+// Finds the end index (exclusive) of the span at `startIndex`, or `endIndex` if
+// there are no other spans between `startIndex` and `endIndex`.
+// The InlineSpan protocol doesn't expose the length of the span so we'll
+// have to iterate through the whole range.
+(InlineSpan, int)? _findEndOfSpan(InlineSpan rootSpan, int startIndex, int endIndex) {
+  assert(endIndex > startIndex);
+  final InlineSpan? subspan = rootSpan.getSpanForPosition(TextPosition(offset: startIndex));
+  if (subspan == null) {
+    return null;
+  }
+  int i = startIndex + 1;
+  while (i < endIndex && rootSpan.getSpanForPosition(TextPosition(offset: i)) == subspan) {
+    i += 1;
+  }
+  return (subspan, i);
+}
+
+/// Helper function that returns 3 special Offsets within the given `box`, to
+/// try performing hit-testing at.
+Iterable<Offset> _testOffsetsForEachTextBox(TextBox box) {
```

Comment:
Could we just hit test the center, similar to https://main-api.flutter.dev/flutter/flutter_test/WidgetController/tap.html?

If we do this, let's also mention it in the docs.

Reply:
I think it's a bit hard to explain how the glyphs are grouped into boxes, if we want to mention hit testing the center of TextBoxes? If the specified font doesn't have glyphs for some characters in the substring then these glyphs will usually be in separate TextBoxes. The 3rd paragraph says the substring must not be obstructed or offscreen, does that work?